kusto/cloudinfo.go

```diff
@@ -0,0 +1,115 @@
+package kusto
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"sync"
+	"time"
+)
+
+// abstraction to query metadata and use this information for providing all
+// information needed for connection string builder to provide all the requisite information
+
+const (
+	metadataPath                  = "v1/rest/auth/metadata"
+	defaultAuthEnvVarName         = "AadAuthorityUri"
+	defaultKustoClientAppId       = "db662dc1-0cfe-4e1c-a843-19a68e65be58"
+	defaultPublicLoginUrl         = "https://login.microsoftonline.com"
+	defaultRedirectUri            = "https://microsoft/kustoclient"
+	defaultKustoServiceResourceId = "https://kusto.kusto.windows.net"
+	defaultFirstPartyAuthorityUrl = "https://login.microsoftonline.com/f8cdef31-a31e-4b4a-93e4-5f571e91255a"
+)
+
+// retrieved metadata
+type metaResp struct {
+	AzureAD CloudInfo
+}
+
+type CloudInfo struct {
+	LoginEndpoint          string `json:"LoginEndpoint"`
+	LoginMfaRequired       bool   `json:"LoginMfaRequired"`
+	KustoClientAppID       string `json:"KustoClientAppId"`
+	KustoClientRedirectURI string `json:"KustoClientRedirectUri"`
+	KustoServiceResourceID string `json:"KustoServiceResourceId"`
+	FirstPartyAuthorityURL string `json:"FirstPartyAuthorityUrl"`
+}
+
+var defaultCloudInfo = CloudInfo{
+	LoginEndpoint:          getEnvOrDefault(defaultAuthEnvVarName, defaultPublicLoginUrl),
+	LoginMfaRequired:       false,
+	KustoClientAppID:       defaultKustoClientAppId,
+	KustoClientRedirectURI: defaultRedirectUri,
+	KustoServiceResourceID: defaultKustoServiceResourceId,
+	FirstPartyAuthorityURL: defaultFirstPartyAuthorityUrl,
+}
+
+// cache to query it once per instance
+var cloudInfoCache sync.Map
+
+func GetMetadata(ctx context.Context, kustoUri string) (CloudInfo, error) {
+	// retrieve &return if exists
+	clusterCachedCloudInfo, ok := cloudInfoCache.Load(kustoUri)
```

Comment:
Doesn't this mean this if two clients will get here at the same time, both will do all of the work?